### ISSUE TITLE
Get predictions hidden states 

### ISSUE THREAD

**Issue by yftah89** [lvapeab/nmt-keras#31]
How can I get the last vectors before the softmax in decoding time? 
more precisly the Ui vectors in this figure - https://raw.githubusercontent.com/lvapeab/nmt-keras/master/examples/documentation/attention_nmt_model.png?token=AEf6E5RhGVqGRSmYi87EbtiGZK7lPxrFks5ZAx-KwA%3D%3D

I want to predict on some test set and get each example's correspinding Ui's
I will prefer to get the ones which corresponding to the best beam search, but I will be glad to get them for beamsearch = 0, meaning for no beam search.

**Comment by lvapeab** [comment]
Hello @yftah89,

this will require some code modifications. As you want to modify the decoder model(s), I think you'll only need to change the sampling models from [in model_zoo.py](https://github.com/lvapeab/nmt-keras/blob/f6d4ddabeb715d1325bd1a5c30e17c7de0522c26/model_zoo.py#L546). 
The vectors you want are stored in [out_layer](https://github.com/lvapeab/nmt-keras/blob/f6d4ddabeb715d1325bd1a5c30e17c7de0522c26/model_zoo.py#L635).
Therefore, you need a model whose outputs are also these vectors. That means, to modify the 
And do something similar in the [model_init](https://github.com/lvapeab/nmt-keras/blob/master/model_zoo.py#L559) and [model_next](https://github.com/lvapeab/nmt-keras/blob/master/model_zoo.py#L653-L654) outputs. Something like:

```
self.model_init = Model(inputs=model_init_input, outputs=model_init_output + [out_layer]) 
```

Now, you'll have a model that outputs your desired vectors. Let's go to how to retrieve them when decoding with beam search. 

___

These changes should be done in the [BeamSearchEnsemble](https://github.com/lvapeab/multimodal_keras_wrapper/blob/master/keras_wrapper/model_ensemble.py#L15) class from the [Multimodal Keras Wrapper](https://github.com/lvapeab/multimodal_keras_wrapper). If you want to output such vectors in training time, the same modifications should be applied in [cnn_model.py](https://github.com/lvapeab/multimodal_keras_wrapper/blob/master/keras_wrapper/cnn_model.py).

So, you'll need to get, in addition to the softmax probabilities, u_i vectors. This can be done in [line#51](https://github.com/lvapeab/multimodal_keras_wrapper/blob/master/keras_wrapper/model_ensemble.py#L51). Something like:
```
[model_probs, next_outs],  u_i = model.predict_cond_optimized(X, states_below, params,
                                                                        ii, prev_out=prev_outs[i])
```
Then, you can append your u_i to another vector (as done with the `alphas_list` and `alphas`). Similarly, you should treat these u_i vectors in the same way as the `alphas` in the [beam_search](https://github.com/lvapeab/multimodal_keras_wrapper/blob/master/keras_wrapper/model_ensemble.py#L71) method.


Good luck!

___

PS: I didn't test this. I probably missed things or some of my suggestions should be modified. But I think this is the way to go. If you need more advice, don't hesitate to ask me :)

PS2: All these instructions assume that you'll set the flag `OPTIMIZED_SEARCH` to `True`.

**Comment by yftah89** [comment]
_First of all let me thank you for your elaborated answer :)_

_Its looks like you were going to write something and the rest of the sentence deleted somehow in "That means, to modify the...", were you trying to write some additional instructions ?_

And do something similar in the model_init and model_next outputs. Something like:
self.model_init = Model(inputs=model_init_input, outputs=model_init_output + [out_layer]) 
_Do I need to add the "+ [out_layer]" in model_next as well?_

**Comment by lvapeab** [comment]
Oops! My fault!  :sweat_smile:
I wrote something different and I didn't edited it correctly. I wanted to say model_init and model_next. I've just edited the previous message.

_Do I need to add the "+ [out_layer]" in model_next as well?_

Yes. The `model_init` computes the first timestep and the `model_next` computes the remaining ones. So, you should also include it there.

**Comment by yftah89** [comment]
It took some time but I mannaged to do it, thanks for your help :)

**Comment by yftah89** [comment]
I also commented in the topic "Some pre-trained weird warning" but it is still in the "close" section so I will be greatful if you take a look :)

**Comment by lvapeab** [comment]
Nice job! I close this one and comment in the other issue.
